File: src/lib/quiz/quiz-engine.ts
import defaultSpec from '../../../content/quiz/quiz.json';

export interface QuizOption {
  id: string;
  label: string;
  score?: number;
}

export interface QuizQuestion {
  id: string;
  dimension: string | null;
  question: string;
  options: QuizOption[];
}

export interface QuizRecommendation {
  title: string;
  action: string;
  refinement: string;
  resource: string;
}

export interface QuizBand {
  min: number;
  max: number;
  id: 'base' | 'consolidar' | 'afinar';
  title: string;
  summary: string;
}

export interface QuizSpec {
  version: string;
  locale: string;
  scoreMeaning: string;
  displayNumericScore: boolean;
  questions: QuizQuestion[];
  priorityOrder: string[];
  recommendations: Record<string, QuizRecommendation>;
  bands: QuizBand[];
  disclaimer: string;
}

export interface QuizPriority {
  dimension: string;
  title: string;
  action: string;
  kind: 'refinement' | 'first_step';
  source: string;
  evidenceAnswer: string;
  resourceId: string;
}

export interface QuizResult {
  source: 'quiz';
  version: string;
  profileId: 'base' | 'consolidar' | 'afinar';
  title: string;
  summary: string;
  internalPreparationScore: number;
  showNumericScore: false;
  context: Record<string, string>;
  dimensions: Record<string, number>;
  priorities: QuizPriority[];
  disclaimer: string;
}

/**
 * Avalia as respostas do questionário seguindo a referência determinística do projeto.
 * Rejeita respostas desconhecidas, incompletas, com tipos inválidos ou opções inexistentes.
 */
export function evaluateQuiz(
  answers: unknown,
  spec: QuizSpec = defaultSpec as unknown as QuizSpec
): QuizResult {
  if (typeof answers !== 'object' || answers === null || Array.isArray(answers)) {
    throw new Error('As respostas devem ser um objeto de escolhas únicas.');
  }

  const ans = answers as Record<string, unknown>;
  const requiredQuestions = new Set(spec.questions.map((q) => q.id));
  const providedKeys = new Set(Object.keys(ans));

  if (
    providedKeys.size !== requiredQuestions.size ||
    [...requiredQuestions].some((id) => !providedKeys.has(id))
  ) {
    throw new Error('O questionário deve conter exatamente as cinco respostas esperadas.');
  }

  const dimensions: Record<string, number> = {};
  const context: Record<string, string> = {};
  const selected: Record<string, string> = {};

  for (const question of spec.questions) {
    const value = ans[question.id];
    if (typeof value !== 'string') {
      throw new Error('Cada pergunta exige uma escolha única válida.');
    }

    const choice = question.options.find((o) => o.id === value);
    if (!choice) {
      throw new Error(`Opção inválida para ${question.id}`);
    }

    selected[question.id] = choice.label;

    if (question.dimension === null) {
      context[question.id] = value;
    } else {
      dimensions[question.dimension] = choice.score ?? 0;
    }
  }

  const total = Object.values(dimensions).reduce((acc, curr) => acc + curr, 0);

  const band = spec.bands.find((b) => total >= b.min && total <= b.max);
  if (!band) {
    throw new Error(`Pontuação ${total} fora de qualquer faixa definida.`);
  }

  const rank: Record<string, number> = {};
  spec.priorityOrder.forEach((name, i) => {
    rank[name] = i;
  });

  const ordered = Object.keys(dimensions).sort((a, b) => {
    const scoreDiff = dimensions[a] - dimensions[b];
    if (scoreDiff !== 0) return scoreDiff;
    return rank[a] - rank[b];
  });

  const priorities: QuizPriority[] = ordered.slice(0, 3).map((dim) => {
    const item = spec.recommendations[dim];
    const isRefinement = dimensions[dim] >= 2;
    const question = spec.questions.find((q) => q.dimension === dim)!;

    return {
      dimension: dim,
      title: item.title,
      action: isRefinement ? item.refinement : item.action,
      kind: isRefinement ? 'refinement' : 'first_step',
      source: 'self_report',
      evidenceAnswer: selected[question.id],
      resourceId: item.resource,
    };
  });

  return {
    source: 'quiz',
    version: spec.version,
    profileId: band.id,
    title: band.title,
    summary: band.summary,
    internalPreparationScore: total,
    showNumericScore: false,
    context,
    dimensions,
    priorities,
    disclaimer: spec.disclaimer,
  };
}

File: reference/quiz_engine.py
"""Reference implementation; port behaviour to TypeScript for the application."""
from __future__ import annotations
import json
from pathlib import Path
from typing import Any

SPEC_PATH = Path(__file__).resolve().parents[1] / 'content/quiz/quiz.json'

def evaluate_quiz(answers: dict[str, str], spec: dict[str, Any] | None = None) -> dict[str, Any]:
    if spec is None:
        spec = json.loads(SPEC_PATH.read_text(encoding='utf-8'))
    if not isinstance(answers, dict):
        raise ValueError('As respostas devem ser um objeto de escolhas únicas.')
    required = {q['id'] for q in spec['questions']}
    if set(answers) != required:
        raise ValueError('O questionário deve conter exatamente as cinco respostas esperadas.')
    dimensions: dict[str, int] = {}
    context: dict[str, str] = {}
    selected: dict[str, str] = {}
    for question in spec['questions']:
        value = answers[question['id']]
        if not isinstance(value, str):
            raise ValueError('Cada pergunta exige uma escolha única válida.')
        choice = next((o for o in question['options'] if o['id'] == value), None)
        if choice is None:
            raise ValueError('Opção inválida para ' + question['id'])
        selected[question['id']] = choice['label']
        dimension = question['dimension']
        if dimension is None:
            context[question['id']] = value
        else:
            dimensions[dimension] = choice['score']
    total = sum(dimensions.values())
    band = next(b for b in spec['bands'] if b['min'] <= total <= b['max'])
    rank = {name: i for i, name in enumerate(spec['priorityOrder'])}
    ordered = sorted(dimensions, key=lambda name: (dimensions[name], rank[name]))
    priorities = []
    for dimension in ordered[:3]:
        item = spec['recommendations'][dimension]
        is_refinement = dimensions[dimension] >= 2
        question = next(q for q in spec['questions'] if q['dimension'] == dimension)
        priorities.append({
            'dimension': dimension,
            'title': item['title'],
            'action': item['refinement'] if is_refinement else item['action'],
            'kind': 'refinement' if is_refinement else 'first_step',
            'source': 'self_report',
            'evidenceAnswer': selected[question['id']],
            'resourceId': item['resource']
        })
    return {
        'source': 'quiz', 'version': spec['version'],
        'profileId': band['id'], 'title': band['title'], 'summary': band['summary'],
        'internalPreparationScore': total, 'showNumericScore': False,
        'context': context, 'dimensions': dimensions, 'priorities': priorities,
        'disclaimer': spec['disclaimer']
    }
